File: golang/tests/pkg/integration/mobile_simulate_integration_test.go
package integration

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/models"
)

const (
	MobileInfoModelA = "SM-A505F"
	MobileInfoModelB = "moto g stylus 5G - 2024"
)

func getInsecureClient(t *testing.T, apiKey string) *agentbay.AgentBay {
	client, err := agentbay.NewAgentBay(apiKey)
	if err != nil {
		t.Fatalf("Failed to create AgentBay client: %v", err)
	}
	return client
}

// TestMobileSimulateForModelAIntegration test mobile simulate feature by model_a prop file
// and check product model is "SM-A505F" after session created
func TestMobileSimulateForModelAIntegration(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping integration test in short mode")
	}

	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("AGENTBAY_API_KEY not set, skipping integration test")
	}

	client := getInsecureClient(t, apiKey)

	t.Log("Upload mobile dev info file for model A...")
	// Assuming the test is running from golang/tests/pkg/integration
	mobileInfoFilePath := filepath.Join("..", "..", "..", "..", "resource", "mobile_info_model_a.json")
	mobileInfoContent, err := os.ReadFile(mobileInfoFilePath)
	if err != nil {
		// Try alternative path if running from root
		mobileInfoFilePath = filepath.Join("resource", "mobile_info_model_a.json")
		mobileInfoContent, err = os.ReadFile(mobileInfoFilePath)
		if err != nil {
			t.Fatalf("Failed to read mobile info file: %v", err)
		}
	}

	// Create mobile simulate service and set simulate params
	t.Log("Creating mobile simulate service and set simulate params...")
	simulateService, err := agentbay.NewMobileSimulateService(client)
	if err != nil {
		t.Fatalf("Failed to create mobile simulate service: %v", err)
	}
	simulateService.SetSimulateEnable(true)
	simulateService.SetSimulateMode(models.MobileSimulateModePropertiesOnly)

	uploadResult := simulateService.UploadMobileInfo(string(mobileInfoContent), nil)
	if !uploadResult.Success {
		t.Fatalf("Failed to upload mobile dev info file: %s", uploadResult.ErrorMessage)
	}
	if uploadResult.MobileSimulateContextID == "" {
		t.Fatal("Expected non-empty MobileSimulateContextID")
	}
	mobileSimContextID := uploadResult.MobileSimulateContextID
	t.Logf("Mobile dev info uploaded successfully: %s", mobileSimContextID)

	params := agentbay.NewCreateSessionParams().
		WithImageId("mobile_latest").
		WithExtraConfigs(&models.ExtraConfigs{
			Mobile: &models.MobileExtraConfig{
				SimulateConfig: simulateService.GetSimulateConfig(),
			},
		})

	result, err := client.Create(params)
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	if !result.Success {
		t.Fatal("Create session returned Success=false")
	}
	if result.Session == nil {
		t.Fatal("Session is nil")
	}
	session := result.Session
	t.Logf("Session created successfully: %s", session.GetSessionId())

	defer func() {
		t.Log("Deleting session...")
		deleteResult, err := client.Delete(session)
		if err != nil {
			t.Errorf("Failed to delete session: %v", err)
		}
		if deleteResult != nil && deleteResult.Success {
			t.Logf("Session deleted successfully (RequestID: %s)", deleteResult.RequestID)
		}
	}()

	time.Sleep(5 * time.Second)
	t.Log("Getting device model after mobile simulate for model A...")
	cmdResult, err := session.GetCommand().ExecuteCommand("getprop ro.product.model")
	if err != nil {
		t.Fatalf("Failed to execute command: %v", err)
	}
	if cmdResult == nil {
		t.Fatal("Command result is nil")
	}

	modelAProductModel := strings.TrimSpace(cmdResult.Output)
	t.Logf("Simulated model A mobile product model: %s", modelAProductModel)
	if modelAProductModel != MobileInfoModelA {
		t.Errorf("Expected device model %s, got %s", MobileInfoModelA, modelAProductModel)
	}
}

// TestMobileSimulateForModelAIntegration test mobile simulate feature by model_b prop file
// and check product model is "moto g stylus 5G - 2024" after session created
func TestMobileSimulateForModelBIntegration(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping integration test in short mode")
	}

	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("AGENTBAY_API_KEY not set, skipping integration test")
	}

	client := getInsecureClient(t, apiKey)

	t.Log("Upload mobile dev info file for model B...")
	mobileInfoFilePath := filepath.Join("..", "..", "..", "..", "resource", "mobile_info_model_b.json")
	mobileInfoContent, err := os.ReadFile(mobileInfoFilePath)
	if err != nil {
		// Try alternative path if running from root
		mobileInfoFilePath = filepath.Join("resource", "mobile_info_model_b.json")
		mobileInfoContent, err = os.ReadFile(mobileInfoFilePath)
		if err != nil {
			t.Fatalf("Failed to read mobile info file: %v", err)
		}
	}

	// Create mobile simulate service and set simulate params
	t.Log("Creating mobile simulate service and set simulate params...")
	simulateService, err := agentbay.NewMobileSimulateService(client)
	if err != nil {
		t.Fatalf("Failed to create mobile simulate service: %v", err)
	}
	simulateService.SetSimulateEnable(true)
	simulateService.SetSimulateMode(models.MobileSimulateModePropertiesOnly)

	uploadResult := simulateService.UploadMobileInfo(string(mobileInfoContent), nil)
	if !uploadResult.Success {
		t.Fatalf("Failed to upload mobile dev info file: %s", uploadResult.ErrorMessage)
	}
	if uploadResult.MobileSimulateContextID == "" {
		t.Fatal("Expected non-empty MobileSimulateContextID")
	}
	mobileSimContextID := uploadResult.MobileSimulateContextID
	t.Logf("Mobile dev info uploaded successfully: %s", mobileSimContextID)

	params := agentbay.NewCreateSessionParams().
		WithImageId("mobile_latest").
		WithExtraConfigs(&models.ExtraConfigs{
			Mobile: &models.MobileExtraConfig{
				SimulateConfig: simulateService.GetSimulateConfig(),
			},
		})

	result, err := client.Create(params)
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	if !result.Success {
		t.Fatal("Create session returned Success=false")
	}
	if result.Session == nil {
		t.Fatal("Session is nil")
	}
	session := result.Session
	t.Logf("Session created successfully: %s", session.GetSessionId())

	defer func() {
		t.Log("Deleting session...")
		deleteResult, err := client.Delete(session)
		if err != nil {
			t.Errorf("Failed to delete session: %v", err)
		}
		if deleteResult != nil && deleteResult.Success {
			t.Logf("Session deleted successfully (RequestID: %s)", deleteResult.RequestID)
		}
	}()

	time.Sleep(5 * time.Second)
	t.Log("Getting device model after mobile simulate for model B...")
	cmdResult, err := session.GetCommand().ExecuteCommand("getprop ro.product.model")
	if err != nil {
		t.Fatalf("Failed to execute command: %v", err)
	}
	if cmdResult == nil {
		t.Fatal("Command result is nil")
	}

	modelBProductModel := strings.TrimSpace(cmdResult.Output)
	t.Logf("Simulated model B mobile product model: %s", modelBProductModel)
	if modelBProductModel != MobileInfoModelB {
		t.Errorf("Expected device model %s, got %s", MobileInfoModelB, modelBProductModel)
	}
}

// TestMobileSimulateForModelAIntegration test mobile simulate persistence feature
// by using a exist mobile simulate context id, and check product model is
// "SM-A505F" after session created
func TestMobileSimulatePersistenceIntegration(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping integration test in short mode")
	}

	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("AGENTBAY_API_KEY not set, skipping integration test")
	}

	client := getInsecureClient(t, apiKey)

	// In Session 1
	t.Log("Upload mobile dev info file for model A...")
	// Assuming the test is running from golang/tests/pkg/integration
	mobileInfoFilePath := filepath.Join("..", "..", "..", "..", "resource", "mobile_info_model_a.json")
	mobileInfoContent, err := os.ReadFile(mobileInfoFilePath)
	if err != nil {
		// Try alternative path if running from root
		mobileInfoFilePath = filepath.Join("resource", "mobile_info_model_a.json")
		mobileInfoContent, err = os.ReadFile(mobileInfoFilePath)
		if err != nil {
			t.Fatalf("Failed to read mobile info file: %v", err)
		}
	}

	// Create mobile simulate service and set simulate params
	t.Log("Creating mobile simulate service and set simulate params...")
	simulateService1, err := agentbay.NewMobileSimulateService(client)
	if err != nil {
		t.Fatalf("Failed to create mobile simulate service: %v", err)
	}
	simulateService1.SetSimulateEnable(true)
	simulateService1.SetSimulateMode(models.MobileSimulateModePropertiesOnly)

	uploadResult := simulateService1.UploadMobileInfo(string(mobileInfoContent), nil)
	if !uploadResult.Success {
		t.Fatalf("Failed to upload mobile dev info file: %s", uploadResult.ErrorMessage)
	}
	if uploadResult.MobileSimulateContextID == "" {
		t.Fatal("Expected non-empty MobileSimulateContextID")
	}
	mobileSimContextID := uploadResult.MobileSimulateContextID
	t.Logf("Mobile dev info uploaded successfully: %s", mobileSimContextID)

	params1 := agentbay.NewCreateSessionParams().
		WithImageId("mobile_latest").
		WithExtraConfigs(&models.ExtraConfigs{
			Mobile: &models.MobileExtraConfig{
				SimulateConfig: simulateService1.GetSimulateConfig(),
			},
		})

	result, err := client.Create(params1)
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	if !result.Success {
		t.Fatal("Create session returned Success=false")
	}
	if result.Session == nil {
		t.Fatal("Session is nil")
	}
	session1 := result.Session
	t.Logf("Session 1 created successfully: %s", session1.GetSessionId())

	defer func() {
		t.Log("Deleting session 1...")
		deleteResult, err := client.Delete(session1)
		if err != nil {
			t.Errorf("Failed to delete session: %v", err)
		}
		if deleteResult != nil && deleteResult.Success {
			t.Logf("Session 1 deleted successfully (RequestID: %s)", deleteResult.RequestID)
		}
	}()

	time.Sleep(5 * time.Second)
	t.Log("Getting device model after mobile simulate for model A...")
	cmdResult, err := session1.GetCommand().ExecuteCommand("getprop ro.product.model")
	if err != nil {
		t.Fatalf("Failed to execute command: %v", err)
	}
	if cmdResult == nil {
		t.Fatal("Command result is nil")
	}

	modelAProductModel := strings.TrimSpace(cmdResult.Output)
	t.Logf("Simulated model A mobile product model: %s", modelAProductModel)
	if modelAProductModel != MobileInfoModelA {
		t.Errorf("Expected device model %s, got %s", MobileInfoModelA, modelAProductModel)
	}

	// In Session 2
	t.Logf("Using a persistent mobild simulate context id: %s", mobileSimContextID)

	// Create mobile simulate service and set simulate params
	t.Log("Creating mobile simulate service and set simulate params...")
	simulateService2, err := agentbay.NewMobileSimulateService(client)
	if err != nil {
		t.Fatalf("Failed to create mobile simulate service: %v", err)
	}
	simulateService2.SetSimulateEnable(true)
	simulateService2.SetSimulateMode(models.MobileSimulateModePropertiesOnly)
	simulateService2.SetSimulateContextID(mobileSimContextID)

	params2 := agentbay.NewCreateSessionParams().
		WithImageId("mobile_latest").
		WithExtraConfigs(&models.ExtraConfigs{
			Mobile: &models.MobileExtraConfig{
				SimulateConfig: simulateService2.GetSimulateConfig(),
			},
		})

	result2, err := client.Create(params2)
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	if !result2.Success {
		t.Fatal("Create session returned Success=false")
	}
	if result2.Session == nil {
		t.Fatal("Session is nil")
	}
	session2 := result2.Session
	t.Logf("Session 2 created successfully: %s", session2.GetSessionId())

	defer func() {
		t.Log("Deleting session 2...")
		deleteResult, err := client.Delete(session2)
		if err != nil {
			t.Errorf("Failed to delete session: %v", err)
		}
		if deleteResult != nil && deleteResult.Success {
			t.Logf("Session 2 deleted successfully (RequestID: %s)", deleteResult.RequestID)
		}
	}()

	time.Sleep(5 * time.Second)
	t.Log("Getting device model after mobile simulate with user context...")
	cmdResult2, err := session2.GetCommand().ExecuteCommand("getprop ro.product.model")
	if err != nil {
		t.Fatalf("Failed to execute command: %v", err)
	}
	if cmdResult2 == nil {
		t.Fatal("Command result is nil")
	}

	productModel2 := strings.TrimSpace(cmdResult2.Output)
	t.Logf("Persistence simulated mobile product model: %s", productModel2)
	if productModel2 != MobileInfoModelA {
		t.Errorf("Expected device model %s, got %s", MobileInfoModelA, productModel2)
	}
}

File: typescript/tests/integration/mobile-simulate.integration.test.ts
/**
 * Integration tests for Mobile Simulate functionality.
 * Tests the complete E2E flow with real API.
 * 
 * Note: These tests create real sessions and should not be run concurrently
 * to avoid rate limiting and resource exhaustion.
 */

import { AgentBay } from '../../src/agent-bay';
import { MobileSimulateService } from '../../src/mobile-simulate';
import { MobileSimulateMode } from '../../src/types/extra-configs';
import { log } from "../../src/utils/logger";
import * as fs from 'fs';
import * as path from 'path';

const MOBILE_INFO_MODEL_A = "SM-A505F";
const MOBILE_INFO_MODEL_B = "moto g stylus 5G - 2024";

describe('Mobile Simulate Integration Tests', () => {
  let agentBay: AgentBay;
  const apiKey = process.env.AGENTBAY_API_KEY;
  let mobileSimPersistenceContextID: string;

  beforeAll(() => {
    if (!apiKey) {
      console.warn('Warning: AGENTBAY_API_KEY not set. Tests will be skipped.');
    }
  });

  beforeEach(() => {
    if (apiKey) {
      agentBay = new AgentBay({ apiKey });
    }
  });

  /**
   * Test mobile simulate feature by model_a prop file
   * and check product model is "SM-A505F" after session created
   */
  describe('Mobile Simulate for Model A', () => {
    test('should simulate Model A mobile device', async () => {
      if (!apiKey) {
        log('Skipping test: AGENTBAY_API_KEY not set');
        return;
      }

      log('Upload mobile dev info file for model A...');
      // Use the service instance from agentBay or create a new one if needed
      // Here we create a new one as per original test, but we could use agentBay.mobileSimulate
      const simulateService = new MobileSimulateService(agentBay);
      simulateService.setSimulateEnable(true);
      simulateService.setSimulateMode(MobileSimulateMode.PropertiesOnly);

      // Read mobile info file
      const mobileInfoPath = path.join(__dirname, '..', '..', '..', 'resource', 'mobile_info_model_a.json');
      const mobileInfoContent = fs.readFileSync(mobileInfoPath, 'utf8');

      // Upload mobile info
      const uploadResult = await simulateService.uploadMobileInfo(mobileInfoContent);
      
      expect(uploadResult.success).toBe(true);
      expect(uploadResult.mobileSimulateContextId).toBeDefined();
      
      const mobileSimContextID = uploadResult.mobileSimulateContextId!;
      mobileSimPersistenceContextID = mobileSimContextID;
      log(`Mobile dev info uploaded successfully: ${mobileSimContextID}`);

      // Create session with mobile simulate
      const sessionResult = await agentBay.create({
        imageId: 'mobile_latest',
        extraConfigs: {
          mobile: {
            lockResolution: false,
            hideNavigationBar: false,
            simulateConfig: simulateService.getSimulateConfig()
          } as any
        }
      });
      
      // Wait, I need to add simulateConfig to MobileExtraConfig interface!
      
      expect(sessionResult.success).toBe(true);
      expect(sessionResult.session).toBeDefined();
      
      const session = sessionResult.session!;
      log(`Session created successfully: ${session.sessionId}`);

      try {
        // Wait for mobile simulate to complete
        await new Promise(resolve => setTimeout(resolve, 5000));

        // Get device model after mobile simulate
        log('Getting device model after mobile simulate for model A...');
        const cmdResult = await (session as any).command.executeCommand('getprop ro.product.model');
        
        expect(cmdResult).toBeDefined();
        expect(cmdResult.success).toBe(true);
        
        const modelAProductModel = cmdResult.output.trim();
        log(`Simulated model A mobile product model: ${modelAProductModel}`);
        
        expect(modelAProductModel).toBe(MOBILE_INFO_MODEL_A);
      } finally {
        // Cleanup
        log('Deleting session...');
        const deleteResult = await session.delete();
        expect(deleteResult.success).toBe(true);
        log(`Session deleted successfully (RequestID: ${deleteResult.requestId})`);
      }
    }, 60000); // 60 second timeout
  });

  /**
   * Test mobile simulate feature by model_b prop file
   * and check product model is "moto g stylus 5G - 2024" after session created
   */
  describe('Mobile Simulate for Model B', () => {
    test('should simulate Model B mobile device', async () => {
      if (!apiKey) {
        log('Skipping test: AGENTBAY_API_KEY not set');
        return;
      }

      log('Upload mobile dev info file for model B...');

      const simulateService = new MobileSimulateService(agentBay);
      simulateService.setSimulateEnable(true);
      simulateService.setSimulateMode(MobileSimulateMode.PropertiesOnly);
      // Read mobile info file
      const mobileInfoPath = path.join(__dirname, '..', '..', '..', 'resource', 'mobile_info_model_b.json');
      const mobileInfoContent = fs.readFileSync(mobileInfoPath, 'utf8');

      // Upload mobile info
      const uploadResult = await simulateService.uploadMobileInfo(mobileInfoContent);
      
      expect(uploadResult.success).toBe(true);
      expect(uploadResult.mobileSimulateContextId).toBeDefined();
      
      const mobileSimContextID = uploadResult.mobileSimulateContextId!;
      log(`Mobile dev info uploaded successfully: ${mobileSimContextID}`);

      // Create session with mobile simulate
      const sessionResult = await agentBay.create({
        imageId: 'mobile_latest',
        extraConfigs: {
          mobile: {
            lockResolution: false,
            hideNavigationBar: false,
            simulateConfig: simulateService.getSimulateConfig()
          } as any
        }
      });

      expect(sessionResult.success).toBe(true);
      expect(sessionResult.session).toBeDefined();
      
      const session = sessionResult.session!;
      log(`Session created successfully: ${session.sessionId}`);

      try {
        // Wait for mobile simulate to complete
        await new Promise(resolve => setTimeout(resolve, 5000));

        // Get device model after mobile simulate
        log('Getting device model after mobile simulate for model B...');
        const cmdResult = await (session as any).command.executeCommand('getprop ro.product.model');
        
        expect(cmdResult).toBeDefined();
        expect(cmdResult.success).toBe(true);
        
        const modelBProductModel = cmdResult.output.trim();
        log(`Simulated model B mobile product model: ${modelBProductModel}`);
        
        expect(modelBProductModel).toBe(MOBILE_INFO_MODEL_B);
      } finally {
        // Cleanup
        log('Deleting session...');
        const deleteResult = await session.delete();
        expect(deleteResult.success).toBe(true);
        log(`Session deleted successfully (RequestID: ${deleteResult.requestId})`);
      }
    }, 60000); // 60 second timeout
  });

  /**
   * Test mobile simulate persistence feature
   * by using an existing mobile simulate context id, and check product model is
   * "SM-A505F" after session created
   */
  describe('Mobile Simulate Persistence', () => {
    test('should use persistent mobile simulate context', async () => {
      if (!apiKey) {
        log('Skipping test: AGENTBAY_API_KEY not set');
        return;
      }

      if (!mobileSimPersistenceContextID) {
        log('Skipping test: mobileSimPersistenceContextID not set (run Model A test first)');
        return;
      }

      log(`Using a persistent mobile simulate context id: ${mobileSimPersistenceContextID}`);
      const simulateService = new MobileSimulateService(agentBay);
      simulateService.setSimulateEnable(true);
      simulateService.setSimulateMode(MobileSimulateMode.PropertiesOnly);
      simulateService.setSimulateContextId(mobileSimPersistenceContextID);

      // Create session with persistent mobile simulate context
      const sessionResult = await agentBay.create({
        imageId: 'mobile_latest',
        extraConfigs: {
          mobile: {
            lockResolution: false,
            hideNavigationBar: false,
            simulateConfig: simulateService.getSimulateConfig()
          } as any
        }
      });

      expect(sessionResult.success).toBe(true);
      expect(sessionResult.session).toBeDefined();
      
      const session = sessionResult.session!;
      log(`Session created successfully: ${session.sessionId}`);

      try {
        // Wait for mobile simulate to complete
        await new Promise(resolve => setTimeout(resolve, 5000));

        // Get device model after mobile simulate
        log('Getting device model after mobile simulate with user context...');
        const cmdResult = await (session as any).command.executeCommand('getprop ro.product.model');
        
        expect(cmdResult).toBeDefined();
        expect(cmdResult.success).toBe(true);
        
        const productModel = cmdResult.output.trim();
        log(`Persistence simulated mobile product model: ${productModel}`);
        
        expect(productModel).toBe(MOBILE_INFO_MODEL_A);
      } finally {
        // Cleanup
        log('Deleting session...');
        const deleteResult = await session.delete();
        expect(deleteResult.success).toBe(true);
        log(`Session deleted successfully (RequestID: ${deleteResult.requestId})`);
      }
    }, 60000); // 60 second timeout
  });
});

